# Tables

In [10]:
!~/Desktop/pypy3/bin/pypy3 run_search.py -p 1 -s 1 2 3 4 5 6 7 8 9 10 11


Solving Air Cargo Problem 1 using breadth_first_search...

# Actions   Expansions   Goal Tests   New Nodes
    20          43          56         178    

Plan length: 6  Time elapsed in seconds: 0.03431642599753104
Load(C1, P1, SFO)
Load(C2, P2, JFK)
Fly(P2, JFK, SFO)
Unload(C2, P2, SFO)
Fly(P1, SFO, JFK)
Unload(C1, P1, JFK)


Solving Air Cargo Problem 1 using depth_first_graph_search...

# Actions   Expansions   Goal Tests   New Nodes
    20          21          22          84    

Plan length: 20  Time elapsed in seconds: 0.008318945008795708
Fly(P1, SFO, JFK)
Fly(P2, JFK, SFO)
Load(C2, P1, JFK)
Fly(P1, JFK, SFO)
Fly(P2, SFO, JFK)
Unload(C2, P1, SFO)
Fly(P1, SFO, JFK)
Fly(P2, JFK, SFO)
Load(C2, P2, SFO)
Fly(P1, JFK, SFO)
Load(C1, P2, SFO)
Fly(P2, SFO, JFK)
Fly(P1, SFO, JFK)
Unload(C2, P2, JFK)
Unload(C1, P2, JFK)
Fly(P2, JFK, SFO)
Load(C2, P1, JFK)
Fly(P1, JFK, SFO)
Fly(P2, SFO, JFK)
Unload(C2, P1, SFO)


Solving Air Cargo Problem 1 using uniform_cost_search...

# Actions   Expansi

## Problem 1
| Algorithm                | Expansions | Goal Tests | New Nodes | Plan Length | Time                 |
| :----------------------: | :--------: | :--------: | --------- | :---------: | :------------------: |
| BFS                      |         43 |         56 |       178 |           6 | 0.03431642599753104  |
| DFS                      |         21 |         22 |        84 |          20 | 0.008318945008795708 |
| UCS                      |         60 |         62 |       240 |           6 | 0.03370450300280936  |
| Greedy BFS + Unmet Goals |          7 |          9 |        29 |           6 | 0.002858270992874168 |
| Greedy BFS + Level Sum   |          6 |          8 |        28 |           6 | 0.6484487549896585   |
| Greedy BFS + Max Level   |          6 |          8 |        24 |           6 | 0.21422760700806975  |
| Greedy BFS + Set Level   |          6 |          8 |        28 |           6 | 0.6197410750028212   |
| A* + Unmet Goals         |         50 |         52 |       206 |           6 | 0.023843417991884053 |
| A* + Level Sum           |         28 |         30 |       122 |           6 | 0.4085546870046528   |
| A* + Max Level           |         43 |         45 |       180 |           6 | 0.2789535109914141   |
| A* + Set Level           |         33 |         35 |       138 |           6 | 0.6717383399954997   |

Number of Actions: 20

In [12]:
!~/Desktop/pypy3/bin/pypy3 run_search.py -p 2 -s 1 2 3 4 5 6 7 8 9 10 11


Solving Air Cargo Problem 2 using breadth_first_search...

# Actions   Expansions   Goal Tests   New Nodes
    72         3343        4609       30503   

Plan length: 9  Time elapsed in seconds: 0.47611921399948187
Load(C1, P1, SFO)
Load(C2, P2, JFK)
Load(C3, P3, ATL)
Fly(P2, JFK, SFO)
Unload(C2, P2, SFO)
Fly(P1, SFO, JFK)
Unload(C1, P1, JFK)
Fly(P3, ATL, SFO)
Unload(C3, P3, SFO)


Solving Air Cargo Problem 2 using depth_first_graph_search...

# Actions   Expansions   Goal Tests   New Nodes
    72         624         625         5602   

Plan length: 619  Time elapsed in seconds: 0.5418473809986608
Fly(P3, ATL, SFO)
Fly(P1, SFO, ATL)
Fly(P3, SFO, JFK)
Fly(P1, ATL, JFK)
Fly(P2, JFK, ATL)
Fly(P3, JFK, ATL)
Fly(P2, ATL, SFO)
Fly(P3, ATL, SFO)
Load(C2, P1, JFK)
Fly(P2, SFO, ATL)
Fly(P1, JFK, ATL)
Fly(P2, ATL, JFK)
Fly(P1, ATL, SFO)
Fly(P3, SFO, ATL)
Fly(P1, SFO, JFK)
Load(C3, P3, ATL)
Fly(P3, ATL, SFO)
Fly(P2, JFK, ATL)
Fly(P3, SFO, JFK)
Fly(P2, ATL, SFO)
Fly(P1, JFK, ATL)
Fly(P2, SFO, J


# Actions   Expansions   Goal Tests   New Nodes
    72         5154        5156       46618   

Plan length: 9  Time elapsed in seconds: 0.9850388869963353
Load(C3, P3, ATL)
Fly(P3, ATL, SFO)
Load(C1, P3, SFO)
Load(C2, P2, JFK)
Fly(P2, JFK, SFO)
Unload(C3, P3, SFO)
Fly(P3, SFO, JFK)
Unload(C2, P2, SFO)
Unload(C1, P3, JFK)


Solving Air Cargo Problem 2 using greedy_best_first_graph_search with h_unmet_goals...

# Actions   Expansions   Goal Tests   New Nodes
    72          17          19         170    

Plan length: 9  Time elapsed in seconds: 0.017671561989118345
Load(C1, P1, SFO)
Load(C2, P2, JFK)
Load(C3, P3, ATL)
Fly(P2, JFK, SFO)
Unload(C2, P2, SFO)
Fly(P3, ATL, SFO)
Unload(C3, P3, SFO)
Fly(P1, SFO, JFK)
Unload(C1, P1, JFK)


Solving Air Cargo Problem 2 using greedy_best_first_graph_search with h_pg_levelsum...

# Actions   Expansions   Goal Tests   New Nodes
    72          9           11          86    

Plan length: 9  Time elapsed in seconds: 2.337402289995225
Load(C1, P1, S

## Problem 2
| Algorithm                | Expansions | Goal Tests | New Nodes | Plan Length | Time                 |
| :----------------------: | :--------: | :--------: | --------- | :---------: | :------------------: |
| BFS                      |       3343 |       4609 |     30503 |           9 | 0.47611921399948187  |
| DFS                      |        624 |        625 |      5602 |         619 | 0.5418473809986608   |
| UCS                      |       5154 |       5156 |     46618 |           9 | 0.9850388869963353   |
| Greedy BFS + Unmet Goals |         17 |         19 |       170 |           9 | 0.017671561989118345 |
| Greedy BFS + Level Sum   |          9 |         11 |        86 |           9 | 2.337402289995225    |
| Greedy BFS + Max Level   |         27 |         29 |       249 |           9 | 1.4095564500021283   |
| Greedy BFS + Set Level   |          9 |         11 |        84 |           9 | 4.637639300999581    |
| A* + Unmet Goals         |       2467 |       2469 |     22522 |           9 | 0.8312939769966761   |
| A* + Level Sum           |        357 |        359 |      3426 |           9 | 22.635809125000378   |
| A* + Max Level           |       2887 |       2889 |     26594 |           9 | 102.79496680898592   |
| A* + Set Level           |       1037 |       1039 |      9605 |           9 | 204.9513987340033    |

Number of Actions: 72

## Problem 3, 4

Out of the three uninformed searches (BFS, DFS, UCS), UCS expands too much, and DFS does not guarantee optimality, so we only use BFS.

For Greedy Best First Search, the the heuristic did not make too large of an impact on Problems 1 and 2. Since Max Level and Set Level heuristics took the most time in Problem 2, we exclude these two heuristics.

Finally, A* search was amazingly slow with its computations. Due to lack of computing resources, I excluded Max Level and Set Level heuristics which took too long.

In [13]:
!~/Desktop/pypy3/bin/pypy3 run_search.py -p 3 -s 1 4 5 8 9


Solving Air Cargo Problem 3 using breadth_first_search...

# Actions   Expansions   Goal Tests   New Nodes
    88        14663       18098       129625  

Plan length: 12  Time elapsed in seconds: 1.0882725579867838
Load(C1, P1, SFO)
Load(C2, P2, JFK)
Fly(P2, JFK, ORD)
Load(C4, P2, ORD)
Fly(P1, SFO, ATL)
Load(C3, P1, ATL)
Fly(P1, ATL, JFK)
Unload(C1, P1, JFK)
Unload(C3, P1, JFK)
Fly(P2, ORD, SFO)
Unload(C2, P2, SFO)
Unload(C4, P2, SFO)


Solving Air Cargo Problem 3 using depth_first_graph_search...

# Actions   Expansions   Goal Tests   New Nodes
    88         408         409         3364   

Plan length: 392  Time elapsed in seconds: 0.26751809900451917
Fly(P1, SFO, ORD)
Fly(P2, JFK, ORD)
Fly(P1, ORD, ATL)
Fly(P2, ORD, ATL)
Fly(P1, ATL, JFK)
Fly(P2, ATL, SFO)
Load(C2, P1, JFK)
Fly(P2, SFO, ORD)
Fly(P1, JFK, ORD)
Fly(P2, ORD, ATL)
Fly(P1, ORD, ATL)
Fly(P2, ATL, JFK)
Fly(P1, ATL, SFO)
Unload(C2, P1, SFO)
Fly(P1, SFO, ORD)
Fly(P2, JFK, ORD)
Fly(P1, ORD, ATL)
Fly(P2, ORD, ATL)
Fly(P1, A


# Actions   Expansions   Goal Tests   New Nodes
    88          14          16         126    

Plan length: 14  Time elapsed in seconds: 2.6600254740042146
Load(C1, P1, SFO)
Fly(P1, SFO, ORD)
Load(C2, P2, JFK)
Fly(P2, JFK, SFO)
Unload(C2, P2, SFO)
Fly(P2, SFO, ATL)
Load(C3, P2, ATL)
Fly(P2, ATL, JFK)
Unload(C3, P2, JFK)
Load(C4, P1, ORD)
Fly(P1, ORD, SFO)
Unload(C4, P1, SFO)
Fly(P1, SFO, JFK)
Unload(C1, P1, JFK)


Solving Air Cargo Problem 3 using astar_search with h_unmet_goals...

# Actions   Expansions   Goal Tests   New Nodes
    88         7388        7390       65711   

Plan length: 12  Time elapsed in seconds: 1.3859413839963963
Load(C2, P2, JFK)
Fly(P2, JFK, ATL)
Load(C3, P2, ATL)
Fly(P2, ATL, ORD)
Load(C4, P2, ORD)
Fly(P2, ORD, SFO)
Unload(C4, P2, SFO)
Unload(C2, P2, SFO)
Load(C1, P2, SFO)
Fly(P2, SFO, JFK)
Unload(C3, P2, JFK)
Unload(C1, P2, JFK)


Solving Air Cargo Problem 3 using astar_search with h_pg_levelsum...

# Actions   Expansions   Goal Tests   New Nodes
    88   

In [16]:
!~/Desktop/pypy3/bin/pypy3 run_search.py -p 4 -s 4 5 8 9


Solving Air Cargo Problem 4 using greedy_best_first_graph_search with h_unmet_goals...

# Actions   Expansions   Goal Tests   New Nodes
   104          29          31         280    

Plan length: 18  Time elapsed in seconds: 0.09087725100107491
Load(C1, P1, SFO)
Load(C2, P2, JFK)
Fly(P2, JFK, SFO)
Unload(C2, P2, SFO)
Fly(P2, SFO, ORD)
Load(C4, P2, ORD)
Load(C5, P2, ORD)
Fly(P2, ORD, SFO)
Unload(C4, P2, SFO)
Fly(P2, SFO, JFK)
Unload(C5, P2, JFK)
Fly(P2, JFK, ATL)
Load(C3, P2, ATL)
Fly(P2, ATL, JFK)
Unload(C3, P2, JFK)
Fly(P2, JFK, SFO)
Fly(P1, SFO, JFK)
Unload(C1, P1, JFK)


Solving Air Cargo Problem 4 using greedy_best_first_graph_search with h_pg_levelsum...

# Actions   Expansions   Goal Tests   New Nodes
   104          17          19         165    

Plan length: 17  Time elapsed in seconds: 5.050209687004099
Fly(P2, JFK, ORD)
Load(C1, P1, SFO)
Fly(P1, SFO, JFK)
Unload(C1, P1, JFK)
Load(C2, P1, JFK)
Fly(P1, JFK, SFO)
Unload(C2, P1, SFO)
Fly(P1, SFO, ATL)
Load(C3, P1, ATL)
Fly(P1,

## Problem 3
| Algorithm                | Expansions | Goal Tests | New Nodes | Plan Length | Time                 |
| :----------------------: | :--------: | :--------: | --------- | :---------: | :------------------: |
| BFS                      |      14633 |      18098 |    129625 |          12 | 1.0882725579867838   |
| Greedy BFS + Unmet Goals |         25 |         27 |       230 |          15 | 0.023119977995520458 |
| Greedy BFS + Level Sum   |         14 |         16 |       126 |          14 | 2.6600254740042146   |
| A* + Unmet Goals         |       7388 |       7390 |     65711 |          12 | 1.3859413839963963   |
| A* + Level Sum           |        369 |        371 |      3403 |          12 | 30.778201508990605   |

Number of Actions: 88

## Problem 4
| Algorithm                | Expansions | Goal Tests | New Nodes | Plan Length | Time                 |
| :----------------------: | :--------: | :--------: | --------- | :---------: | :------------------: |
| BFS                      |      99736 |     114953 |    944130 |          14 | 6.470810332000838    |
| Greedy BFS + Unmet Goals |         29 |         31 |       280 |          18 | 0.09087725100107491  |
| Greedy BFS + Level Sum   |         17 |         16 |       165 |          17 | 5.050209687004099    |
| A* + Unmet Goals         |      34330 |      34332 |    328509 |          14 | 7.168496578000486    |
| A* + Level Sum           |       1208 |       1210 |     12210 |          15 | 202.31166815999313   |

Number of Actions: 104

# Charts

In [ ]:
import matplotlib.pyplot as plt

## Expanded Nodes vs Number of Actions

# Questions

## Which algorithm or algorithms would be most appropriate for planning in a very restricted domain (i.e., one that has only a few actions) and needs to operate in real time?

## Which algorithm or algorithms would be most appropriate for planning in very large domains (e.g., planning delivery routes for all UPS drivers in the U.S. on a given day)

The Greedy Best First Search with the unmet goals heuristic will perform very well in all aspects every when the number of actions is large.

## Which algorithm or algorithms would be most appropriate for planning problems where it is important to find only optimal plans?

For uninformed search, BFS and UCS are guaranteed to find the optimal plan, but not DFS.

For informed search, the heuristic must be admissible to guarantee optimality. 